src/Java.Interop.Export/Java.Interop/MarshalMemberBuilder.cs

```diff
@@ -120,11 +120,15 @@ public LambdaExpression CreateMarshalToManagedExpression (MethodInfo method, Jav
 
 			var envp        = Expression.Variable (typeof (JniTransition), "__envp");
 			var jvm         = Expression.Variable (typeof (JniRuntime), "__jvm");
+			var vm          = Expression.Variable (typeof (JniRuntime.JniValueManager), "__vm");
 			var envpVars    = new List<ParameterExpression> () {
 				envp,
 				jvm,
 			};
 
+			if (!method.IsStatic)
+				envpVars.Add (vm);
```

Comment:
I don't see why we don't use `vm` for static methods, especially considering that static methods *also* call `__jvm.ValueManager.WaitForGCBridgeProcessing()`, so having an extra local variable for `__jvm.ValueManager` costs us nothing.

Additionally, if any *parameters* require marshaling, we'll *likewise* potentially require the value marshaler, so by *not* including this we may generate invalid IL.

This also shows a problem in our existing unit tests: the existing tests, e.g. [`CreateMarshalFromJniMethodExpression_StaticFuncMyLegacyColorMyColor_MyColor `](https://github.com/xamarin/java.interop/blob/de3bc3bd840c498b81576c3d7ca8f023932b842b/src/Java.Interop.Export/Tests/Java.Interop/MarshalMemberBuilderTest.cs#L285-L324), *only* use `ValueManager.GetValue<T>()` for the `__this` parameter, and no others.

This suggests that we need to add a method to `ExportTest` which will result in using `ValueManager.GetValue<T>()` for a parameter, on both an instance and `static` method, to ensure we don't break this scenario.

Reply:
> I don't see why we don't use vm for static methods, especially considering that static methods also call __jvm.ValueManager.WaitForGCBridgeProcessing(), so having an extra local variable for __jvm.ValueManager costs us nothing.

For static methods without parameters it produces shorter methods, so it would cost us a bit of space.

It is true that when there are marshaled parameters we should still use the variable, so I will change  the condition to also check the marshaled parameters count.

Reply:
It *might* cost *some* space, but not much, and always adding the temporary makes it easier for us to understand and maintain.

...and I do mean *might*: it depends on compiler optimizations. Consider this terrible sample app:

```csharp
class ValueManager {
	public void WaitForGCBridgeProcessing() {}
}
class JniRuntime {
	public ValueManager ValueManager => new ValueManager ();
	public static JniRuntime CurrentRuntime => new JniRuntime ();
}

class Example {
	static void WithTemporary()
	{
		var r = JniRuntime.CurrentRuntime;
		var v = r.ValueManager;
		v.WaitForGCBridgeProcessing();
	}
	static void WithoutTemporary()
	{
		var r = JniRuntime.CurrentRuntime;
		r.ValueManager.WaitForGCBridgeProcessing ();
	}
}

class App {
	public static void Main ()
	{
	}
}
```

build *with* debugging:

```
$ csc /debug+ /optimize- pr305.cs
```

...and `Example.WithTemporary()` is 22 bytes of code, while `Example.WithoutTemporary()` is 20 bytes of code.

Build with optimizations:

```
$ csc /debug- /optimize+ pr305.cs
```

and *both* `Example.WithTemporary()` *and* `Example.WithoutTemporary()` are *identical*, 16 bytes in size.

I don't think we need to worry about the extra code size that the `__vm` variable implies.

Reply:
I already updated the code in the evening to have better condition for vm variable inclusion.

The condition is reasonably simple and the resulting code even looks better IMO, when we don't introduce variable which is used only once.

About the savings, I measured them on Release version of Mono.Android.dll and using the vm variable conditionally indeed helps, while having it always nearly vanish the savings.

| version                           | size              | saving |
|------------------------|-------------|--------|
| vm variable conditional | 38635008   | 18944 |
| vm variable always        | 38652928   | 1024   |
| original                           | 38653952  |             |

I think the reason here is that LINQ expressions are not much optimized, so there it makes a difference.

For the compiled examples you tried, it would be interesting to see whether the sizes remain the same when you add a catch block we have in out methods, which uses the runtime variable.

```
catch (Exception ex) when (r.ExceptionShouldTransitionToJni (ex)) {
		jniTransition.SetPendingException (ex);
		return false;
	}
```